packages/angular_devkit/build_optimizer/src/purify/purify.ts

```diff
@@ -5,38 +5,99 @@
  * Use of this source code is governed by an MIT-style license that can be
  * found in the LICENSE file at https://angular.io/license
  */
+import { RawSource, ReplaceSource } from 'webpack-sources';
+
+
 // This matches a comment left by the build-optimizer that contains pure import paths
 const importCommentRegex = /\/\*\* PURE_IMPORTS_START (\S+) PURE_IMPORTS_END \*\//mg;
 
-// TODO: handle sourcemaps
-export function purify(content: string) {
+// Replacements are meant to be used with Webpack's ReplaceSource.
+export interface Replacement {
+  start: number;
+  end: number;
+  content: string;
+}
+
+export function purifyReplacements(content: string) {
 
   const pureImportMatches = getMatches(content, importCommentRegex, 1).join('|');
+  const replacements: Replacement[] = [];
+
+  /* Prefix safe imports with pure */
+  content.replace(
+    new RegExp(`(_(${pureImportMatches})__ = )(__webpack_require__\\(\\S+\\);)`, 'mg'),
+    (match, p1, _p2, p3, offset) => {
+      const newContent = `${p1}/*@__PURE__*/${p3}`;
+      replacements.push({
+        start: offset,
+        end: offset + match.length - 1,
+        content: newContent,
+      });
+
+      return newContent;
+    },
+  );
+
+  /* Prefix default safe imports with pure */
+  content.replace(
+    new RegExp(
+      `(_(${pureImportMatches})___default = )(__webpack_require__\\.\\w\\(\\S+\\);)`, 'mg',
+    ),
+    (match, p1, _p2, p3, offset) => {
+      const newContent = `${p1}/*@__PURE__*/${p3}`;
+      replacements.push({
+        start: offset,
+        end: offset + match.length - 1,
+        content: newContent,
+      });
+
+      return newContent;
+    },
+  );
+
+  /* Prefix CCF and CMF statements */
+  content.replace(
+    /\w*__WEBPACK_IMPORTED_MODULE_\d+__angular_core__\["\w+" \/\* (ɵccf|ɵcmf|ɵcrt) \*\/\]\(/mg,
+    (match, _p1, offset) => {
+      const newContent = `/*@__PURE__*/${match}`;
+      replacements.push({
+        start: offset,
+        end: offset + match.length - 1,
+        content: newContent,
+      });
+
+      return newContent;
+    },
+  );
+
+  /* Prefix module statements */
+  content.replace(
+    /new \w*__WEBPACK_IMPORTED_MODULE_\d+__angular_core__\["\w+" \/\* NgModuleFactory \*\/\]/mg,
+    (match, offset) => {
+      const newContent = `/*@__PURE__*/${match}`;
+      replacements.push({
+        start: offset,
+        end: offset + match.length - 1,
+        content: newContent,
+      });
```

Comment:
Nit: Lines 48 - 54 are pretty similar to 63 - 69 and 78-84. Could we re-use any of the code in a shared function?

Reply:
They are indeed, fixed.